### PR TITLE
Fix #3277: Created BUILD.bazel for domain/state

### LINKED ISSUE #3277: Create a BUILD.bazel file for domain/state
This task is part of the build system migration from Gradle to Bazel. 

## Prerequesites

- Make sure you are able to build Oppia using Bazel
- Follow the [Oppia Bazel Setup instructions](https://github.com/oppia/oppia-android/wiki/Oppia-Bazel-Setup-Instructions) to setup Bazel and learn the terminology.
- Read the [Gradle Bazel Migration Best Practies and FAQ](https://github.com/oppia/oppia-android/wiki/Gradle--Bazel-Migration-Best-Practices-and-FAQ) for more context and see if there's an answer there to any question you have.

## Task

1. Create a BUILD.bazel file in

`domain/src/main/java/org/oppia/android/domain/state/`

1. Define the kt_android_library library/libraries

Based on the imports in the Kotlin files, add the required third_party dependencies to make the libraries build. Check the Useful links section for a list of third_party dependencies.  You may also need to add the non third_party dependencies.

This package should only depend on libraries from the model module.

3. Add the directory to the MIGRATED_PROD_FILES of the module's root BUILD.bazel file

4. Add the libraries to the list of deps dependencies in the targets that require the library

You will have to set the apropriate visbility. See the list of visibilities in the link below.

5. Clean up any unused deps dependencies

6. Verify that the app builds

## Useful links

- [BUILD.bazel files](https://github.com/oppia/oppia-android/search?q=filename%3ABUILD.bazel&type=code)
- [Visibilities](https://github.com/oppia/oppia-android/blob/76dc13dba3ad77aecd2218a5372949d6551f12fe/BUILD.bazel)
- [third_party dependencies](https://github.com/oppia/oppia-android/blob/a70d9b016f0d77dd30c4939041f82ea02ee28d1d/third_party/versions.bzl)

### PR DESCRIPTION
**Pull request by yash10019coder**
## Explanation

Fixes #3277 
Created a BUILD.bazel file in domain state
## Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

### CODE REVIEW

**Comment by yash10019coder** [comment]
@fsharpasharp @anandwana001 @prayutsu  created a PR but unable to add dependencies in BUILD.bazel please help

**Comment by yash10019coder** [comment]
All the dependencies in the domain/utility classes are of model and it has no bazel packages for it therefore I'm haveing trouble including the dependencies of the domain/utility classes

**Comment by BenHenning** [comment]
@yash10019coder you'll need to add dependencies from https://github.com/oppia/oppia-android/blob/develop/model/BUILD.bazel, e.g.: ``//model:exploration_java_proto_lite`` for each of the protos that are used by these classes. I think Exploration gives you most of them, but you might need a few more.

**Comment by yash10019coder** [comment]
> @yash10019coder you'll need to add dependencies from https://github.com/oppia/oppia-android/blob/develop/model/BUILD.bazel, e.g.: `//model:exploration_java_proto_lite` for each of the protos that are used by these classes. I think Exploration gives you most of them, but you might need a few more.

@BenHenning I have added these dependencies in my latest commit but I still get some errors and how did you found out that we have to put this in dependencies //model:exploration_java_proto_lite three is no mention of the java files in it

and I'm attaching the screenshot of the error

Screenshot from 2021-08-21 10-13-22

**Comment by prayutsu** [comment]
> > @yash10019coder you'll need to add dependencies from https://github.com/oppia/oppia-android/blob/develop/model/BUILD.bazel, e.g.: `//model:exploration_java_proto_lite` for each of the protos that are used by these classes. I think Exploration gives you most of them, but you might need a few more.
> 
> @BenHenning I have added these dependencies in my latest commit but I still get some errors and how did you found out that we have to put this in dependencies //model:exploration_java_proto_lite three is no mention of the java files in it
> 
> and I'm attaching the screenshot of the error
> 
> Screenshot from 2021-08-21 10-13-22

@yash10019coder To find out what deps need to be added, you can check the imports of the file `StateDeck.kt` as imports tell us what other classes are we using in that file.
If you look at the error carefully, it says `unresolved reference` which means that Bazel does not know about those particular classes when it tries to build these newly introduces libraries. If you want to tell Bazel about these classes, then you need to add those classes in the `deps` as well.

File: .idea/runConfigurations.xml [hunk 1]
```diff
@@ -1,12 +0,0 @@
...
-  <component name="RunConfigurationProducerService">
-    <option name="ignoredProducers">
-      <set>
-        <option value="org.jetbrains.plugins.gradle.execution.test.runner.AllInPackageGradleConfigurationProducer" />
-        <option value="org.jetbrains.plugins.gradle.execution.test.runner.TestClassGradleConfigurationProducer" />
-        <option value="org.jetbrains.plugins.gradle.execution.test.runner.TestMethodGradleConfigurationProducer" />
-      </set>
-    </option>
-  </component>
-</project>
```
Comment (prayutsu) [inline .idea/runConfigurations.xml]: Please restore this file.

File: domain/src/main/java/org/oppia/android/domain/state/BUILD.bazel [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+kt_android_library(
+    name = "state_deck",
+    srcs = [
+        "StateDeck.kt",
+    ],
+    visibility = ["//:oppia_api_visibility"],
+    deps = [
+        "//model:exploration_java_proto_lite"
+    ]
```
Comment (prayutsu) [inline domain/src/main/java/org/oppia/android/domain/state/BUILD.bazel]: Add all missing deps that are being highlighted in the build error. 
Ditto for other added libraries.

**Review by prayutsu — CHANGES_REQUESTED** [review]
@yash10019coder Left some comments, PTAL.

File: domain/src/main/java/org/oppia/android/domain/state/BUILD.bazel (hunk 2 above)
Comment (yash10019coder) [reply domain/src/main/java/org/oppia/android/domain/state/BUILD.bazel]: @prayutsu  that I have understood but the main problem is how to find which proto is to be added
like AnswersAndResponse.java is being used in StateDeck but now how can I find the particular of this in bazel/model

Comment (prayutsu) [reply domain/src/main/java/org/oppia/android/domain/state/BUILD.bazel]: You need to find the proto file that contains the message `AnswerAndResponse` in it and then you can specify the library of that file here.
To find the .proto file just press `Ctrl + Shift + R` and type the keywords that you want to search in the whole codebase.

**Review by prayutsu — COMMENTED** [review]
@yash10019coder Followed up on our comment.

**Comment by yash10019coder** [comment]
@prayutsu @fsharpasharp made a new commit PTAL 
I'm still getting errors althought I have added all the dependencies 
attaching the SS below

Screenshot from 2021-08-22 10-10-23
Screenshot from 2021-08-22 10-10-37
Screenshot from 2021-08-22 10-10-51
Screenshot from 2021-08-22 10-11-12
Screenshot from 2021-08-22 10-11-21
Screenshot from 2021-08-22 10-11-28

**Comment by yash10019coder** [comment]
@prayutsu @fsharpasharp @BenHenning PTAL

**Comment by yash10019coder** [comment]
Made a commit @prayutsu @fsharpasharp @BenHenning 
Still getting the error PTAL and here's the SS
Screenshot from 2021-08-23 12-11-58
Screenshot from 2021-08-23 12-14-46

File: domain/src/main/java/org/oppia/android/domain/state/BUILD.bazel [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    visibility = ["//:oppia_api_visibility"],
+    deps = [
+        "//model:exploration_checkpoint_java_proto_lite",
+        "//model:exploration_java_proto_lite",
+        "//model:feedback_reporting_java_proto_lite",
+        "//model:interaction_object_java_proto_lite",
+        "//model:onboarding_java_proto_lite",
+        "//model:question_java_proto_lite",
+        "//model:subtitled_html_java_proto_lite",
+        "//model:topic_java_proto_lite",
```
Comment (BenHenning) [inline domain/src/main/java/org/oppia/android/domain/state/BUILD.bazel]: We shouldn't be adding all of these--you should look at the import statements within StateDeck to see which model dependencies are needed, then add them here & build locally to verify.

Ditto for the other new libraries below.

File: .idea/runConfigurations.xml (hunk 1 above)
Comment (BenHenning) [reply .idea/runConfigurations.xml]: It appears this file was removed again--please restore it.

File: domain/src/main/java/org/oppia/android/domain/state/BUILD.bazel [hunk 4]
```diff
@@ -0,0 +1,57 @@
+"""
+build bazel for domain state
```
Comment (BenHenning) [inline domain/src/main/java/org/oppia/android/domain/state/BUILD.bazel]: Suggest instead:

```
Domain services & definitions corresponding to managing ephemeral states in play sessions.
```

File: domain/src/main/java/org/oppia/android/domain/state/StateDeck.kt [hunk 5]
```diff
@@ -85,7 +85,7 @@ internal class StateDeck internal constructor(
   }
 
   /** Returns the current [EphemeralState] the learner is viewing. */
-  internal fun getCurrentEphemeralState(helpIndex: HelpIndex): EphemeralState {
```
Comment (BenHenning) [inline domain/src/main/java/org/oppia/android/domain/state/StateDeck.kt:88]: Did you mean to remove this parameter? It seems like it's still needed.

File: domain/src/main/java/org/oppia/android/domain/state/StateDeck.kt [hunk 6]
```diff
@@ -128,14 +128,53 @@ internal class StateDeck internal constructor(
...
+    val ephemeralState = EphemeralState.newBuilder()
+      .setState(newState)
+      .setHasPreviousState(!isCurrentStateInitial())
+      .setPendingState(
+        PendingState.newBuilder().addAllWrongAnswer(currentDialogInteractions).addAllHint(hintList)
+      )
+      .build()
+    pendingTopState = newState
+    solutionIsRevealed = true
+    return ephemeralState
```
Comment (BenHenning) [inline domain/src/main/java/org/oppia/android/domain/state/StateDeck.kt]: Did these come from a merge conflict? I don't think we want to reintroduce this code.

In general, please make sure to double check your changes before pushing to ensure exactly the changes you intend to submit are pushed.

Ditto elsewhere in this & the other changed files.

File: model/BUILD.bazel [hunk 7]
```diff
@@ -178,6 +179,7 @@ format_import_proto_library(
 
 java_lite_proto_library(
     name = "feedback_reporting_java_proto_lite",
+    visibility = ["//:oppia_api_visibility"],
```
Comment (BenHenning) [inline model/BUILD.bazel]: Here & above: I suspect these visibility changes aren't actually needed.

**Review by BenHenning — CHANGES_REQUESTED** [review]
Thanks @yash10019coder. I think a bunch of extra changes got pulled in, maybe from a recent merge?

PTAL at latest comments.

**Comment by BenHenning** [comment]
@yash10019coder the deps you added should result in no build failures (though there is extra code that's been added). Can you try addressing all the comments in this PR & then build again to see if the problem gets addressed?

File: domain/src/main/java/org/oppia/android/domain/state/StateDeck.kt (hunk 5 above)
Comment (yash10019coder) [reply domain/src/main/java/org/oppia/android/domain/state/StateDeck.kt:88]: Actually I didn't mean to remove the param I wanted to remove the internal keyword it was removed by mistake

File: domain/src/main/java/org/oppia/android/domain/state/StateDeck.kt (hunk 6 above)
Comment (yash10019coder) [reply domain/src/main/java/org/oppia/android/domain/state/StateDeck.kt]: Yes there was a merge confilct

Comment (yash10019coder) [reply domain/src/main/java/org/oppia/android/domain/state/StateDeck.kt]: Shall I remove this or let it be ?

**Comment by yash10019coder** [comment]
@prayutsu @BenHenning made a PR PTAL

**Comment by BenHenning** [comment]
Hi @yash10019coder. My apologies for the delay--I'll need to take a look at this tomorrow.

**Comment by BenHenning** [comment]
Hi--sorry again about the delay. I still plan to look at this later in the week; I've just had some project emergencies to deal with in the meantime. Thanks for your patience.

**Comment by yash10019coder** [comment]
@BenHenning please review the PR it is marked staled

File: .idea/runConfigurations.xml [hunk 8]
```diff
@@ -3,9 +3,12 @@
...
       <set>
-        <option value="org.jetbrains.plugins.gradle.execution.test.runner.AllInPackageGradleConfigurationProducer" />
-        <option value="org.jetbrains.plugins.gradle.execution.test.runner.TestClassGradleConfigurationProducer" />
-        <option value="org.jetbrains.plugins.gradle.execution.test.runner.TestMethodGradleConfigurationProducer" />
+        <option
+          value="org.jetbrains.plugins.gradle.execution.test.runner.AllInPackageGradleConfigurationProducer" />
+        <option
+          value="org.jetbrains.plugins.gradle.execution.test.runner.TestClassGradleConfigurationProducer" />
+        <option
+          value="org.jetbrains.plugins.gradle.execution.test.runner.TestMethodGradleConfigurationProducer" />
```
Comment (BenHenning) [inline .idea/runConfigurations.xml]: Please revert these changes--they're unrelated to the PR.

**Review by BenHenning — COMMENTED** [review]
Thanks @yash10019coder. So sorry for the extended delay--I will prioritize reviewing this this week.

I had one new comment, and I think one of my earlier comments hasn't yet been resolved. Please make sure to address & reply to all comments before sending a PR back to be re-reviewed.

**Review by yash10019coder — COMMENTED** [review]
@BenHenning resolved the comments

**Comment by yash10019coder** [comment]
@BenHenning PTAL

File: domain/src/main/java/org/oppia/android/domain/state/StateDeck.kt (hunk 6 above)
Comment (BenHenning) [reply domain/src/main/java/org/oppia/android/domain/state/StateDeck.kt]: I think you've fixed it in the latest change. All I see are the removals of 'internal' which seems correct.

File: domain/src/main/java/org/oppia/android/domain/state/BUILD.bazel [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+
+kt_android_library(
+    name = "state_list",
+    srcs = [
+        "StateList.kt",
+    ],
+    visibility = ["//:oppia_api_visibility"],
+    deps = [
+        "//model:event_logger_java_proto_lite",
+        "//model:exploration_checkpoint_java_proto_lite",
```
Comment (BenHenning) [inline domain/src/main/java/org/oppia/android/domain/state/BUILD.bazel]: Are these two needed?

**Review by BenHenning — COMMENTED** [review]
Thanks @yash10019coder. Just 2 new comments.

Similar to my comment in your other PR, please make sure to respond via a comment to all conversation threads from reviewers to confirm you saw them & addressed/didn't address them.

File: domain/src/main/java/org/oppia/android/domain/state/BUILD.bazel (hunk 9 above)
Comment (yash10019coder) [reply domain/src/main/java/org/oppia/android/domain/state/BUILD.bazel]: Yes I think it is needed

Comment (yash10019coder) [reply domain/src/main/java/org/oppia/android/domain/state/BUILD.bazel]: import org.oppia.android.app.model.Outcome
because of this import on line 4 in StateList.kt

Comment (yash10019coder) [reply domain/src/main/java/org/oppia/android/domain/state/BUILD.bazel]: import org.oppia.android.app.model.State
on line 6 in StateList.kt

Comment (BenHenning) [reply domain/src/main/java/org/oppia/android/domain/state/BUILD.bazel]: We should be using exploration_java_proto_lite, then, for those (not the checkpoint version of it).

**Review by BenHenning — COMMENTED** [review]
Thanks @yash10019coder. Justhad one follow-up.

**Comment by yash10019coder** [comment]
@BenHenning removed unused and unncesscarty imports from the bazel build file PTAL

**Review by BenHenning — APPROVED** [review]
Thanks @yash10019coder. This LGTM!

**Comment by BenHenning** [comment]
@prayutsu it seems you requested changes. Could you PTAL?

**Review by prayutsu — APPROVED** [review]
LGTM, Thanks @yash10019coder

**Comment by prayutsu** [comment]
@BenHenning This is ready to merge now 😄

**Comment by BenHenning** [comment]
Thanks @prayutsu & @yash10019coder.
